### PR TITLE
AppDomain safety

### PR DESCRIPTION
**Pull request by BrennanConroy**
@Tratcher @davidfowl @kichalla 

Related to https://github.com/aspnet/Caching/issues/203

### CODE REVIEW

File: test/Microsoft.Extensions.Logging.Test/ConsoleLoggerTest.cs [hunk 1]
```diff
@@ -642,6 +642,38 @@ public void CallingBeginScopeOnLogger_ReturnsNonNullableInstance()
...
+
+            // Act
+            domain.DoCallBack(DomainFunc);
+            AppDomain.Unload(domain);
+            var disposable = logger.BeginScope("Scope1");
+            logger.LogInformation("Test");
+
+            // Assert
+            Assert.NotNull(disposable);
+            Assert.Equal(4, sink.Writes.Count);
```
Comment (kichalla) [inline test/Microsoft.Extensions.Logging.Test/ConsoleLoggerTest.cs]: I can't recall how these writes look like...how is the count 4 here? (for example, the DomainFunc creates a new logger and new sink, so we are including it's numbers right?)

Comment (kichalla) [inline test/Microsoft.Extensions.Logging.Test/ConsoleLoggerTest.cs]: Ah scratch that. Got it.

**Comment by BrennanConroy** [comment]
@kichalla Is it good?

**Comment by kichalla** [comment]
:shipit:

**Pull request merged**